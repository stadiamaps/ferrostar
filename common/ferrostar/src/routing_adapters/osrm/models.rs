//! OSRM models from the API spec: <http://project-osrm.org/docs/v5.5.1/api/>
//!
//! Note that in some cases we optionally allow for extensions that have been made to the spec
//! by others which are now pseudo-standardized (ex: Mapbox). We omit some fields which are not
//! needed for navigation.

use crate::models::{ManeuverModifier, ManeuverType};
use alloc::{string::String, vec::Vec};
use serde::Deserialize;
use serde_json::Value;
#[cfg(feature = "alloc")]
use std::collections::HashMap;

#[derive(Deserialize, Debug)]
#[serde(transparent)]
pub struct Coordinate {
    tuple: (f64, f64),
}

impl Coordinate {
    pub fn latitude(&self) -> f64 {
        self.tuple.1
    }

    pub fn longitude(&self) -> f64 {
        self.tuple.0
    }
}

#[derive(Deserialize, Debug)]
pub struct RouteResponse {
    /// The response code.
    ///
    /// Ok indicates success. TODO: enumerate others?
    pub code: String,
    pub routes: Vec<Route>,
    pub waypoints: Vec<Waypoint>,
}

/// A route between two or more waypoints.
#[derive(Deserialize, Debug)]
pub struct Route {
    /// The estimated travel time, in seconds.
    pub duration: f64,
    /// The distance traveled by the route, in meters.
    pub distance: f64,
    /// The geometry of the route.
    ///
    /// NOTE: This library assumes that 1) an overview geometry will always be requested, and
    /// 2) that it will be a polyline (whether it is a polyline5 or polyline6 can be determined
    /// by the [`crate::routing_adapters::RouteResponseParser`]).
    pub geometry: String,
    /// The legs between the given waypoints.
    pub legs: Vec<RouteLeg>,
}

/// A route between exactly two waypoints.
#[derive(Deserialize, Debug)]
pub struct RouteLeg {
    pub annotation: Option<AnyAnnotation>,
    /// The estimated travel time, in seconds.
    pub duration: f64,
    /// The distance traveled this leg, in meters.
    pub distance: f64,
    /// A sequence of steps with turn-by-turn instructions.
    pub steps: Vec<RouteStep>,
    /// A Mapbox and Valhalla extension which indicates which waypoints are passed through rather than creating a new leg.
    #[serde(default)]
    pub via_waypoints: Vec<ViaWaypoint>,
}

#[derive(Deserialize, Debug, Clone, PartialEq)]
pub struct AnyAnnotation {
    #[serde(flatten)]
    pub values: HashMap<String, Vec<Value>>,
}

/// The max speed json units that can be associated with annotations.
/// [MaxSpeed/Units](https://wiki.openstreetmap.org/wiki/Map_Features/Units#Speed)
/// TODO: We could generalize this as or map from this to a `UnitSpeed` enum.
#[derive(Deserialize, PartialEq, Debug, Clone)]
pub enum MaxSpeedUnits {
    #[serde(rename = "km/h")]
    KilometersPerHour,
    #[serde(rename = "mph")]
    MilesPerHour,
}

/// The local posted speed limit between a pair of coordinates.
#[derive(Debug, Clone)]
#[allow(dead_code)] // TODO: https://github.com/stadiamaps/ferrostar/issues/271
pub enum MaxSpeed {
    Unknown { unknown: bool },
    Known { speed: f64, unit: MaxSpeedUnits },
}

#[allow(dead_code)] // TODO: https://github.com/stadiamaps/ferrostar/issues/271
impl MaxSpeed {
    /// Get the max speed as meters per second.
    pub fn get_as_meters_per_second(&self) -> Option<f64> {
        match self {
            MaxSpeed::Known { speed, unit } => match unit {
                MaxSpeedUnits::KilometersPerHour => Some(speed * 0.27778),
                MaxSpeedUnits::MilesPerHour => Some(speed * 0.44704),
            },
            #[allow(unused)]
            MaxSpeed::Unknown { unknown } => None,
        }
    }
}

#[derive(Deserialize, Debug)]
pub struct RouteStep {
    /// The distance from the start of the current maneuver to the following step, in meters.
    pub distance: f64,
    /// The estimated travel time, in seconds.
    pub duration: f64,
    /// The (unsimplified) geometry of the route segment.
    ///
    /// NOTE: This library assumes that the geometry will always be a polyline.
    pub geometry: String,
    /// The name of the way along which travel proceeds.
    pub name: Option<String>,
    /// A reference number or code for the way (if one is available).
    #[serde(rename = "ref")]
    pub reference: Option<String>,
    /// A pronunciation hint for the name of the way.
    pub pronunciation: Option<String>,
    /// The mode of transportation.
    pub mode: Option<String>,
    /// The maneuver for this step
    pub maneuver: StepManeuver,
    /// TODO: docs
    pub intersections: Vec<Intersections>,

    /// A list of exits (name or number), separated by semicolons.
    ///
    /// NOTE: This annotation is not in the official spec, but is a common extension used by Mapbox
    /// and Valhalla.
    pub exits: Option<String>,

    /// The side of the way on which traffic proceeds.
    ///
    /// NOTE: This annotation is not in the official spec, but is a common extension used by Mapbox
    /// and Valhalla.
    pub driving_side: Option<String>,
    // Mapbox and Valhalla extensions that might be useful later
    // pub rotary_name: Option<String>,
    // pub rotary_pronunciation: Option<String>,
    /// Textual instructions that are displayed as a banner; supported by Mapbox and Valhalla
    #[serde(default, rename = "bannerInstructions")]
    pub banner_instructions: Vec<BannerInstruction>,
    /// Textual instructions that are displayed as a banner; supported by Mapbox and Stadia Maps
    #[serde(default, rename = "voiceInstructions")]
    pub voice_instructions: Vec<VoiceInstruction>,
}

#[derive(Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct BannerInstruction {
    /// How far (in meters) from the upcoming maneuver the instruction should start being displayed
    pub distance_along_geometry: f64,
    pub primary: BannerContent,
    pub secondary: Option<BannerContent>,
}

#[derive(Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct VoiceInstruction {
    pub announcement: String,
    pub ssml_announcement: Option<String>,
    /// How far (in meters) from the upcoming maneuver the instruction should be announced
    pub distance_along_geometry: f64,
}

#[derive(Deserialize, Debug)]
pub struct BannerContent {
    pub text: String,
    #[serde(rename = "type")]
    pub maneuver_type: Option<ManeuverType>,
    #[serde(rename = "modifier")]
    pub maneuver_modifier: Option<ManeuverModifier>,
    /// The degree at which the maneuver exits the roundabout.
    /// 180 indicates that the effect of exiting will be to continue in the same direction as
    /// the original travel.
    #[serde(rename = "degrees")]
    pub roundabout_exit_degrees: Option<u16>,
}

#[derive(Deserialize, Debug)]
pub struct StepManeuver {
    /// The coordinate at which the maneuver takes place.
    pub location: Coordinate,
    /// The clockwise angle from true north to the direction of travel immediately *before*
    /// the maneuver.
    pub bearing_before: u16,
    /// The clockwise angle from true north to the direction of travel immediately *after*
    /// the maneuver.
    pub bearing_after: u16,
    /// A string indicating the type of maneuver.
    ///
    /// Note that even though there are `new name` and `notification` instructions, the
    /// `mode` and `name` (of the parent [`RouteStep`]) can change between *any* pair of instructions.
    /// They only offer a fallback in case there is nothing else to report.
    /// TODO: Model this as an enum. Note that new types may be introduced, and anything unknown to the client should be handled like a turn.
    #[serde(rename = "type")]
    pub maneuver_type: String,
    /// An optional string indicating the direction change of the maneuver.
    /// TODO: Model this as an enum.
    pub modifier: Option<String>,
    /// Non-standard extension in Mapbox and Valhalla where the instruction is computed server-side
    instruction: Option<String>,
}

impl StepManeuver {
    // TODO: This is a placeholder implementation.
    // Most commercial offerings offer server-side synthesis of voice instructions.
    // However, we might consider synthesizing these locally too.
    // This will be rather cumbersome with localization though.
    fn synthesize_instruction(&self, _locale: &str) -> String {
        String::from("TODO: OSRM instruction synthesis")
    }

    pub fn get_instruction(&self) -> String {
        self.instruction
            .clone()
            .unwrap_or_else(|| self.synthesize_instruction("en-US"))
    }
}

#[derive(Deserialize, Debug)]
pub struct Intersections {
    /// The location of the intersection
    pub location: Coordinate,
    /// A list of bearing values that are available at the intersection.
    ///
    /// These describe all available roads at the intersection.
    pub bearings: Vec<u16>,
    /// An array of strings signifying the classes (as specified in the profile) of the road
    /// exiting the intersection.
    ///
    /// Note that Valhalla servers do not return this property.
    #[serde(default)]
    pub classes: Vec<String>,
    /// A list of entry flags, corresponding 1:1 to the list of bearings.
    ///
    /// This value indicates whether the respective road could be entered on a valid route (not
    /// violating some restriction).
    pub entry: Vec<bool>,
    /// An index into the bearings/entry array used to calculate the bearing just before the turn.
    ///
    /// The clockwise angle from true north to the direction of travel immediately before the
    /// maneuver/passing the intersection.
    /// To get the bearing in the direction of driving, the bearing has to be rotated by
    /// 180 degrees. Not supplied for `depart` maneuvers.
    #[serde(default)]
    #[serde(rename = "in")]
    pub intersection_in: usize,
    /// An index into the bearings/entry array used to calculate the bearing just before the turn.
    ///
    /// The clockwise angle from true north to the direction of travel immediately after the
    /// maneuver/passing the intersection.  Not supplied for `arrive` maneuvers.
    #[serde(default)]
    #[serde(rename = "out")]
    pub intersection_out: usize,
    /// A list of turn [`Lane`]s available at the intersection (if info is available).
    ///
    /// Lanes are listed in left-to-right order.
    #[serde(default)]
    pub lanes: Vec<Lane>,
}

#[derive(Deserialize, Debug)]
pub struct Lane {
    /// An indication (ex: marking on the road, sign, etc.) for a turn lane.
    ///
    /// A lane may have multiple indications (ex: both straight and left),
    /// TODO: Turn this into an enum
    pub indications: Vec<String>,
    /// Whether the lane is a valid choice for the current maneuver
    pub valid: bool,
    // TODO: Mapbox and Valhalla extensions: `active` and `valid_indication`
}

#[derive(Deserialize, Debug)]
pub struct Waypoint {
    /// The name of the street that the waypoint snapped to.
    pub name: Option<String>,
    /// The distance (in meters) between the snapped point and the input coordinate.
    pub distance: Option<f64>,
    /// The waypoint's location on the road network.
    pub location: Coordinate,
}

#[derive(Deserialize, Debug)]
pub struct ViaWaypoint {
    /// The distance (in meters) from the leg origin
    pub distance_from_start: f64,
    /// The geometry point index of the location (leg-specific).
    pub geometry_index: f64,
    /// The waypoint's index in the array of waypoints.
    pub waypoint_index: usize,
}

#[cfg(test)]
mod tests {
    use super::*;

    // TODO: Route
    // TODO: RouteLeg

    #[test]
    fn deserialize_annotation() {
        // Example from Mapbox's public docs, which include several annotations not supported at
        // the time of this writing.
        let data = r#"{
            "distance": [
                4.294596842089401,
                5.051172053200946,
                5.533254065167979,
                6.576513793849532,
                7.4449640160938015,
                8.468757534990829,
                15.202780313562714,
                7.056346577326572
            ],
            "duration": [
                1,
                1.2,
                2,
                1.6,
                1.8,
                2,
                3.6,
                1.7
            ],
            "speed": [
                4.3,
                4.2,
                2.8,
                4.1,
                4.1,
                4.2,
                4.2,
                4.2
            ],
            "congestion": [
                "low",
                "moderate",
                "moderate",
                "moderate",
                "heavy",
                "heavy",
                "heavy",
                "heavy"
            ],
            "maxspeed": [
              {
                "speed": 56,
                "unit": "km/h"
              },
              {
                "speed": 56,
                "unit": "km/h"
              },
              {
                "speed": 56,
                "unit": "km/h"
              },
              {
                "speed": 56,
                "unit": "km/h"
              },
              {
                "speed": 56,
                "unit": "km/h"
              },
              {
                "speed": 56,
                "unit": "km/h"
              },
              {
                "speed": 56,
                "unit": "km/h"
              },
              {
                "speed": 56,
                "unit": "km/h"
              }
            ]
        }"#;

        let annotation: AnyAnnotation =
            serde_json::from_str(data).expect("Failed to parse Annotation");

        insta::with_settings!({sort_maps => true}, {
            insta::assert_yaml_snapshot!(annotation.values);
        });
    }

    #[test]
    fn deserialize_banner_instruction() {
        // Example from Mapbox's public docs
        let data = r#"
        {
          "distanceAlongGeometry": 100,
          "primary": {
            "type": "turn",
            "modifier": "left",
            "text": "I 495 North / I 95",
            "components": [
              {
                "text": "I 495",
                "imageBaseURL": "https://s3.amazonaws.com/mapbox/shields/v3/i-495",
                "type": "icon"
              },
              {
                "text": "North",
                "type": "text",
                "abbr": "N",
                "abbr_priority": 0
              },
              {
                "text": "/",
                "type": "delimiter"
              },
              {
                "text": "I 95",
                "imageBaseURL": "https://s3.amazonaws.com/mapbox/shields/v3/i-95",
                "type": "icon"
              }
            ]
          },
          "secondary": {
            "type": "turn",
            "modifier": "left",
            "text": "Baltimore / Northern Virginia",
            "components": [
              {
                "text": "Baltimore",
                "type": "text"
              },
              {
                "text": "/",
                "type": "text"
              },
              {
                "text": "Northern Virginia",
                "type": "text"
              }
            ]
          },
          "sub": {
            "text": "",
            "components": [
              {
                "text": "",
                "type": "lane",
                "directions": [
                  "left"
                ],
                "active": true
              },
              {
                "text": "",
                "type": "lane",
                "directions": [
                  "left",
                  "straight"
                ],
                "active": true
              },
              {
                "text": "",
                "type": "lane",
                "directions": [
                  "right"
                ],
                "active": false
              }
            ]
          }
        }
        "#;

        let instruction: BannerInstruction =
            serde_json::from_str(data).expect("Failed to parse Annotation");

        assert_eq!(instruction.distance_along_geometry, 100.0);
        assert_eq!(instruction.primary.text, "I 495 North / I 95");
        assert_eq!(instruction.primary.maneuver_type, Some(ManeuverType::Turn));
        assert_eq!(
            instruction.primary.maneuver_modifier,
            Some(ManeuverModifier::Left)
        );

        let secondary = instruction.secondary.expect("Expected secondary content");
        assert_eq!(secondary.text, "Baltimore / Northern Virginia");
        assert_eq!(secondary.maneuver_type, Some(ManeuverType::Turn));
        assert_eq!(secondary.maneuver_modifier, Some(ManeuverModifier::Left));
    }

    #[test]
    fn test_max_speed_unknown() {
        let max_speed = MaxSpeed::Unknown { unknown: true };
        assert_eq!(max_speed.get_as_meters_per_second(), None);
    }

    #[test]
    fn test_max_speed_kph() {
        let max_speed = MaxSpeed::Known {
            speed: 100.0,
            unit: MaxSpeedUnits::KilometersPerHour,
        };
        assert_eq!(
            max_speed.get_as_meters_per_second(),
            Some(27.778000000000002)
        );
    }

    #[test]
    fn test_max_speed_mph() {
        let max_speed = MaxSpeed::Known {
            speed: 60.0,
            unit: MaxSpeedUnits::MilesPerHour,
        };
        assert_eq!(max_speed.get_as_meters_per_second(), Some(26.8224));
    }
}
