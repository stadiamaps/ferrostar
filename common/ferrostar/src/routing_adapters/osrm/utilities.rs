use super::models::AnyAnnotation;
use crate::models::AnyAnnotationValue;
use crate::routing_adapters::error::ParsingError;
use serde_json::Value;
use std::collections::HashMap;

/// Gets a slice of the route's annotations array.
///
/// Returns a [`ParsingError`] if the annotations are not present or the slice is out of bounds.
pub(crate) fn get_annotation_slice(
    annotations: Option<Vec<AnyAnnotationValue>>,
    start_index: usize,
    end_index: usize,
) -> Result<Vec<AnyAnnotationValue>, ParsingError> {
    annotations
        .ok_or(ParsingError::MalformedAnnotations {
            error: "No annotations".to_string(),
        })?
        .get(start_index..end_index)
        .ok_or(ParsingError::MalformedAnnotations {
            error: "Annotations slice index out of bounds ({start_index}..{end_index})".to_string(),
        })
        .map(<[AnyAnnotationValue]>::to_vec)
}

/// Converts the the OSRM-style annotation object consisting of separate arrays
/// to a single vector of parsed objects (one for each coordinate pair).
pub(crate) fn zip_annotations(annotation: AnyAnnotation) -> Vec<AnyAnnotationValue> {
    let source: HashMap<String, Vec<Value>> = annotation.values;

    // Get the length of the array (assumed to be the same for all annotations)
    let length = source.values().next().map_or(0, Vec::len);

    return (0..length)
        .map(|i| {
            source
                .iter()
                .filter_map(|(key, values)| {
                    // Values is the vector at a specific key in the original annotations object.
                    values
                        .get(i) // For each key, get the value at the index i.
                        .map(|v| (key.clone(), v.clone())) // Convert the key and value to a tuple.
                })
                .collect::<HashMap<String, Value>>() // Collect the key-value pairs into a hashmap.
        })
        .map(|value| AnyAnnotationValue { value })
        .collect::<Vec<AnyAnnotationValue>>();
}

#[cfg(test)]
mod test {

    use serde_json::Map;

    use super::*;

    #[test]
    fn test_zip_annotation() {
        let json_str = r#"{
            "distance": [1.2, 2.24, 3.24],
            "duration": [4, 5, 6],
            "speed": [10, 11, 12],
            "max_speed": [{
              "speed": 56,
              "unit": "km/h"
            }, {
              "speed": 12,
              "unit": "mi/h"
            }, {
              "unknown": true
            }],
            "traffic": ["bad", "ok", "good"],
            "construction": [null, true, null]
        }"#;

        let json_value: Map<String, Value> = serde_json::from_str(json_str).unwrap();
        let values: HashMap<String, Vec<Value>> = json_value
            .iter()
            .map(|(k, v)| (k.to_string(), v.as_array().unwrap().clone()))
            .collect();

        // Construct the annotation object.
        let annotation = AnyAnnotation { values };
        let result = zip_annotations(annotation);

        insta::with_settings!({sort_maps => true}, {
            insta::assert_yaml_snapshot!(result);
        });
    }
}
