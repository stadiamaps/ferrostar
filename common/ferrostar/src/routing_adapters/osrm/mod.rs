//! Response parsing for OSRM-compatible JSON (including Stadia Maps, Valhalla, Mapbox, etc.).

pub(crate) mod models;
pub mod utilities;

use super::RouteResponseParser;
use crate::models::{
    AnyAnnotationValue, GeographicCoordinate, Incident, LaneInfo, RouteStep, SpokenInstruction,
    VisualInstruction, VisualInstructionContent, Waypoint, WaypointKind,
};
use crate::routing_adapters::utilities::get_coordinates_from_geometry;
use crate::routing_adapters::{
    osrm::models::{
        Route as OsrmRoute, RouteResponse, RouteStep as OsrmRouteStep, Waypoint as OsrmWaypoint,
    },
    ParsingError, Route,
};
#[cfg(all(not(feature = "std"), feature = "alloc"))]
use alloc::{string::ToString, vec, vec::Vec};
use geo::BoundingRect;
use models::BannerContent;
use polyline::decode_polyline;
use utilities::get_annotation_slice;
use uuid::Uuid;

/// A response parser for OSRM-compatible routing backends.
///
/// The parser is NOT limited to only the standard OSRM format; many Valhalla/Mapbox tags are also
/// parsed and are included in the final route.
#[derive(Debug)]
pub struct OsrmResponseParser {
    polyline_precision: u32,
}

impl OsrmResponseParser {
    pub fn new(polyline_precision: u32) -> Self {
        Self { polyline_precision }
    }
}

impl RouteResponseParser for OsrmResponseParser {
    fn parse_response(&self, response: Vec<u8>) -> Result<Vec<Route>, ParsingError> {
        let res: RouteResponse = serde_json::from_slice(&response)?;

        if res.code == "Ok" {
            res.routes
                .iter()
                .map(|route| Route::from_osrm(route, &res.waypoints, self.polyline_precision))
                .collect::<Result<Vec<_>, _>>()
        } else {
            Err(ParsingError::InvalidStatusCode { code: res.code })
        }
    }
}

impl Route {
    pub fn from_osrm(
        route: &OsrmRoute,
        waypoints: &[OsrmWaypoint],
        polyline_precision: u32,
    ) -> Result<Self, ParsingError> {
        let via_waypoint_indices: Vec<_> = route
            .legs
            .iter()
            .flat_map(|leg| leg.via_waypoints.iter().map(|via| via.waypoint_index))
            .collect();

        let waypoints: Vec<_> = waypoints
            .iter()
            .enumerate()
            .map(|(idx, waypoint)| Waypoint {
                coordinate: GeographicCoordinate {
                    lat: waypoint.location.latitude(),
                    lng: waypoint.location.longitude(),
                },
                kind: if via_waypoint_indices.contains(&idx) {
                    WaypointKind::Via
                } else {
                    WaypointKind::Break
                },
            })
            .collect();

        let linestring = decode_polyline(&route.geometry, polyline_precision).map_err(|error| {
            ParsingError::InvalidGeometry {
                error: error.to_string(),
            }
        })?;
        if let Some(bbox) = linestring.bounding_rect() {
            let geometry: Vec<GeographicCoordinate> = linestring
                .coords()
                .map(|coord| GeographicCoordinate::from(*coord))
                .collect();

            let steps = route
                .legs
                .iter()
                .flat_map(|leg| {
                    // Converts all single value annotation vectors into a single vector witih a value object.
                    let annotations = leg
                        .annotation
                        .as_ref()
                        .map(|leg_annotation| utilities::zip_annotations(leg_annotation.clone()));

                    // Convert all incidents into a vector of Incident objects.
                    let incident_items = leg
                        .incidents
                        .iter()
                        .map(Incident::from)
                        .collect::<Vec<Incident>>();

                    // Index for the annotations slice
                    let mut start_index: usize = 0;

                    leg.steps.iter().map(move |step| {
                        let step_geometry =
                            get_coordinates_from_geometry(&step.geometry, polyline_precision)?;

                        // Slice the annotations for the current step.
                        // The annotations array represents segments between coordinates.
                        //
                        // 1. Annotations should never repeat.
                        // 2. Each step has one less annotation than coordinate.
                        // 3. The last step never has annotations as it's two of the route's last coordinate (duplicate).
                        let step_index_len = step_geometry.len() - 1_usize;
                        let end_index = start_index + step_index_len;

                        let annotation_slice =
                            get_annotation_slice(annotations.clone(), start_index, end_index).ok();

                        let relevant_incidents_slice = incident_items
                            .iter()
                            .filter(|incident| {
                                let incident_start = incident.geometry_index_start as usize;

                                match incident.geometry_index_end {
                                    Some(end) => {
                                        let incident_end = end as usize;
                                        incident_start >= start_index && incident_end <= end_index
                                    }
                                    None => {
                                        incident_start >= start_index && incident_start <= end_index
                                    }
                                }
                            })
                            .map(|incident| {
                                let mut adjusted_incident = incident.clone();
                                if adjusted_incident.geometry_index_start - start_index as u64 > 0 {
                                    adjusted_incident.geometry_index_start -= start_index as u64;
                                } else {
                                    adjusted_incident.geometry_index_start = 0;
                                }

                                if let Some(end) = adjusted_incident.geometry_index_end {
                                    let adjusted_end = end - start_index as u64;
                                    adjusted_incident.geometry_index_end =
                                        Some(if adjusted_end > end_index as u64 {
                                            end_index as u64
                                        } else {
                                            adjusted_end
                                        });
                                }
                                adjusted_incident
                            })
                            .collect::<Vec<Incident>>();

                        start_index = end_index;

                        RouteStep::from_osrm_and_geom(
                            step,
                            step_geometry,
                            annotation_slice,
                            relevant_incidents_slice,
                        )
                    })
                })
                .collect::<Result<Vec<_>, _>>()?;

            Ok(Route {
                geometry,
                bbox: bbox.into(),
                distance: route.distance,
                waypoints: waypoints.clone(),
                steps,
            })
        } else {
            Err(ParsingError::InvalidGeometry {
                error: "Bounding box could not be calculated".to_string(),
            })
        }
    }
}

impl RouteStep {
    fn extract_exit_numbers(banner_content: &BannerContent) -> Vec<String> {
        banner_content
            .components
            .iter()
            .filter(|component| component.component_type.as_deref() == Some("exit-number"))
            .filter_map(|component| component.text.clone())
            .collect()
    }

    fn from_osrm_and_geom(
        value: &OsrmRouteStep,
        geometry: Vec<GeographicCoordinate>,
        annotations: Option<Vec<AnyAnnotationValue>>,
        incidents: Vec<Incident>,
    ) -> Result<Self, ParsingError> {
        let visual_instructions = value
            .banner_instructions
            .iter()
            .map(|banner| VisualInstruction {
                primary_content: VisualInstructionContent {
                    text: banner.primary.text.clone(),
                    maneuver_type: banner.primary.maneuver_type,
                    maneuver_modifier: banner.primary.maneuver_modifier,
                    roundabout_exit_degrees: banner.primary.roundabout_exit_degrees,
                    lane_info: None,
                    exit_numbers: Self::extract_exit_numbers(&banner.primary),
                },
                secondary_content: banner.secondary.as_ref().map(|secondary| {
                    VisualInstructionContent {
                        text: secondary.text.clone(),
                        maneuver_type: secondary.maneuver_type,
                        maneuver_modifier: secondary.maneuver_modifier,
                        roundabout_exit_degrees: banner.primary.roundabout_exit_degrees,
                        lane_info: None,
                        exit_numbers: Self::extract_exit_numbers(&secondary),
                    }
                }),
                sub_content: banner.sub.as_ref().map(|sub| VisualInstructionContent {
                    text: sub.text.clone(),
                    maneuver_type: sub.maneuver_type,
                    maneuver_modifier: sub.maneuver_modifier,
                    roundabout_exit_degrees: sub.roundabout_exit_degrees,
                    lane_info: {
                        let lane_infos: Vec<LaneInfo> = sub
                            .components
                            .iter()
                            .filter(|component| component.component_type.as_deref() == Some("lane"))
                            .map(|component| LaneInfo {
                                active: component.active.unwrap_or(false),
                                directions: component.directions.clone().unwrap_or_default(),
                                active_direction: component.active_direction.clone(),
                            })
                            .collect();

                        if lane_infos.is_empty() {
                            None
                        } else {
                            Some(lane_infos)
                        }
                    },
                    exit_numbers: Self::extract_exit_numbers(&sub),
                }),
                trigger_distance_before_maneuver: banner.distance_along_geometry,
            })
            .collect();

        let spoken_instructions = value
            .voice_instructions
            .iter()
            .map(|instruction| SpokenInstruction {
                text: instruction.announcement.clone(),
                ssml: instruction.ssml_announcement.clone(),
                trigger_distance_before_maneuver: instruction.distance_along_geometry,
                utterance_id: Uuid::new_v4(),
            })
            .collect();

        // Convert the annotations to a vector of json strings.
        // This allows us to safely pass the RouteStep through the FFI boundary.
        // The host platform can then parse an arbitrary annotation object.
        let annotations_as_strings: Option<Vec<String>> = annotations.map(|annotations_vec| {
            annotations_vec
                .iter()
                .map(|annotation| serde_json::to_string(annotation).unwrap())
                .collect()
        });

        let exits = match value.exits.clone() {
            Some(exit_text) => exit_text.split(';').map(|s| s.trim().to_string()).collect(),
            None => Vec::new(),
        };

        Ok(RouteStep {
            geometry,
            // TODO: Investigate using the haversine distance or geodesics to normalize.
            // Valhalla in particular is a bit nonstandard. See https://github.com/valhalla/valhalla/issues/1717
            distance: value.distance,
            duration: value.duration,
            road_name: value.name.clone(),
            exits,
            instruction: value.maneuver.get_instruction(),
            visual_instructions,
            spoken_instructions,
            annotations: annotations_as_strings,
            incidents,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const STANDARD_OSRM_POLYLINE6_RESPONSE: &str =
        include_str!("fixtures/standard_osrm_polyline6_response.json");
    const VALHALLA_OSRM_RESPONSE: &str = include_str!("fixtures/valhalla_osrm_response.json");
    const VALHALLA_OSRM_RESPONSE_VIA_WAYS: &str =
        include_str!("fixtures/valhalla_osrm_response_via_ways.json");
    const VALHALLA_EXTENDED_OSRM_RESPONSE: &str =
        include_str!("fixtures/valhalla_extended_osrm_response.json");
    const VALHALLA_OSRM_RESPONSE_WITH_EXITS: &str =
        include_str!("fixtures/valhalla_osrm_response_with_exit_info.json");

    #[test]
    fn parse_standard_osrm() {
        let parser = OsrmResponseParser::new(6);
        let routes = parser
            .parse_response(STANDARD_OSRM_POLYLINE6_RESPONSE.into())
            .expect("Unable to parse OSRM response");
        insta::assert_yaml_snapshot!(routes);
    }

    #[test]
    fn parse_valhalla_osrm() {
        let parser = OsrmResponseParser::new(6);
        let routes = parser
            .parse_response(VALHALLA_OSRM_RESPONSE.into())
            .expect("Unable to parse Valhalla OSRM response");

        insta::assert_yaml_snapshot!(routes, {
            ".**.annotations" => "redacted annotations json strings vec"
        });
    }

    #[test]
    fn parse_valhalla_osrm_with_via_ways() {
        let parser = OsrmResponseParser::new(6);
        let routes = parser
            .parse_response(VALHALLA_OSRM_RESPONSE_VIA_WAYS.into())
            .expect("Unable to parse Valhalla OSRM response");

        insta::assert_yaml_snapshot!(routes, {
            ".**.annotations" => "redacted annotations json strings vec"
        });
    }

    #[test]
    fn parse_valhalla_asserting_annotation_lengths() {
        let parser = OsrmResponseParser::new(6);
        let routes = parser
            .parse_response(VALHALLA_OSRM_RESPONSE.into())
            .expect("Unable to parse Valhalla OSRM response");

        // Loop through every step and validate that the length of the annotations
        // matches the length of the geometry minus one. This is because each annotation
        // represents a segment between two coordinates.
        for (route_index, route) in routes.iter().enumerate() {
            for (step_index, step) in route.steps.iter().enumerate() {
                if step_index == route.steps.len() - 1 {
                    // The arrival step is 2 of the same coordinates.
                    // So annotations will be None.
                    assert_eq!(step.annotations, None);
                    continue;
                }

                let step = step.clone();
                let annotations = step.annotations.expect("No annotations");
                assert_eq!(
                    annotations.len(),
                    step.geometry.len() - 1,
                    "Route {}, Step {}",
                    route_index,
                    step_index
                );
            }
        }
    }

    #[test]
    fn parse_valhalla_asserting_sub_maneuvers() {
        let parser = OsrmResponseParser::new(6);
        let routes = parser
            .parse_response(VALHALLA_EXTENDED_OSRM_RESPONSE.into())
            .expect("Unable to parse Valhalla Extended OSRM response");

        // Collect all sub_contents into a vector
        let sub_contents: Vec<_> = routes
            .iter()
            .flat_map(|route| &route.steps)
            .filter_map(|step| {
                step.visual_instructions
                    .iter()
                    .find_map(|instruction: &VisualInstruction| instruction.sub_content.as_ref())
            })
            .collect();

        // Assert that there's exactly one sub maneuver instructions as is the case in the test data
        assert_eq!(
            sub_contents.len(),
            1,
            "Expected exactly one sub banner instructions"
        );

        if let Some(sub_content) = sub_contents.first() {
            // Ensure that there are 4 pieces of lane information in the sub banner instructions
            if let Some(lane_info) = &sub_content.lane_info {
                assert_eq!(lane_info.len(), 4);
            } else {
                panic!("Expected lane information, but could not find it");
            }
        } else {
            panic!("No sub banner instructions found in any of the steps")
        }
    }

    #[test]
    fn parse_osrm_with_exits() {
        let parser = OsrmResponseParser::new(6);
        let routes = parser
            .parse_response(VALHALLA_OSRM_RESPONSE_WITH_EXITS.into())
            .expect("Unable to parse OSRM response");

        insta::assert_yaml_snapshot!(routes, {
            ".**.annotations" => "redacted annotations json strings vec"
        });
    }
}
