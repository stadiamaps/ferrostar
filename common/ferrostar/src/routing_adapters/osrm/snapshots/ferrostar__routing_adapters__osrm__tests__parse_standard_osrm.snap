---
source: ferrostar/src/routing_adapters/osrm/mod.rs
expression: routes
---
- geometry:
    - lat: 52.517033
      lng: 13.388798
    - lat: 52.527168
      lng: 13.387228
    - lat: 52.528491
      lng: 13.393668
    - lat: 52.529432
      lng: 13.39763
    - lat: 52.529684
      lng: 13.403888
    - lat: 52.528326
      lng: 13.411389
    - lat: 52.527507
      lng: 13.41432
    - lat: 52.52677
      lng: 13.415657
    - lat: 52.528458
      lng: 13.417166
    - lat: 52.528728
      lng: 13.421348
    - lat: 52.528082
      lng: 13.424085
    - lat: 52.528068
      lng: 13.424993
    - lat: 52.527885
      lng: 13.425184
    - lat: 52.527043
      lng: 13.427263
    - lat: 52.525063
      lng: 13.43036
    - lat: 52.52479
      lng: 13.430413
    - lat: 52.523269
      lng: 13.429678
    - lat: 52.523239
      lng: 13.428554
  bbox:
    sw:
      lat: 52.517033
      lng: 13.387228
    ne:
      lat: 52.529684
      lng: 13.430413
  distance: 4731.8
  waypoints:
    - coordinate:
        lat: 52.517033
        lng: 13.388798
      kind: Break
      properties: "{\"name\":\"Friedrichstraße\",\"distance\":4.231521214}"
    - coordinate:
        lat: 52.529432
        lng: 13.39763
      kind: Break
      properties: "{\"name\":\"Torstraße\",\"distance\":2.795148358}"
    - coordinate:
        lat: 52.523239
        lng: 13.428554
      kind: Break
      properties: "{\"name\":\"Platz der Vereinten Nationen\",\"distance\":2.226580806}"
  steps: []
